core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -280,20 +282,32 @@ class Partition(val topic: String,
             updateIsr(newInSyncReplicas)
             replicaManager.isrExpandRate.mark()
           }
-
+          // check if the LW of the partition can now be incremented
+          // since the replica's LSO may have incremented
+          leaderLWIncremented = maybeIncrementLeaderLW(leaderReplica)
           // check if the HW of the partition can now be incremented
           // since the replica may already be in the ISR and its LEO has just incremented
-          maybeIncrementLeaderHW(leaderReplica, logReadResult.fetchTimeMs)
+          leaderHWIncremented = maybeIncrementLeaderHW(leaderReplica, logReadResult.fetchTimeMs)
 
-        case None => false // nothing to do if no longer leader
+        case None => // nothing to do if no longer leader
       }
     }
 
-    // some delayed operations may be unblocked after HW changed
-    if (leaderHWIncremented)
+    // some delayed operations may be unblocked after HW or LW changed
+    if (leaderLWIncremented || leaderHWIncremented)
       tryCompleteDelayedRequests()
   }
 
+  def checkLowWatermarkReachOffset(requiredOffset: Long): (Boolean, Errors, Long) = {
```

Comment:
Methods with three tuples as return value may be a little hard to follow, may be we can create a case class. At very least we should document each field of the return value.